src/box/lua/net_box.lua

```diff
@@ -386,12 +386,20 @@ local function new_sm(uri, opts)
     -- Now it is necessary to have a strong ref to callback somewhere or
     -- it is GC-ed prematurely. We store a strong reference in the remote
     -- connection object.
+    local transport
     local function weak_callback(...)
         local callback = weak_refs.callback
         if callback then return callback(...) end
+        -- The callback is responsible for handling graceful shutdown.
+        -- If it's garbage collected, the connection won't be closed on
+        -- receiving 'box.shutdown' event and so the server won't exit
+        -- until the connection object is garbage collected, which may
+        -- take forever. To avoid that, let's break the worker loop if
+        -- we see that the callback is unavailable.
+        transport:stop()
```

Comment:
Are you sure you can ref the transport variable inside of `weak_callback`? I made this test:
```Lua
box.cfg{listen = 3313}
box.schema.user.grant('guest', 'super')
net = require('net.box')
c = net.connect(box.cfg.listen)
c:eval('return 100')
c:close()
c = nil
collectgarbage()
collectgarbage()
collectgarbage()
```
and applied this patch:
```Diff
diff --git a/src/box/lua/net_box.c b/src/box/lua/net_box.c
index d79f90fd9..99330aa5f 100644
--- a/src/box/lua/net_box.c
+++ b/src/box/lua/net_box.c
@@ -1727,6 +1727,7 @@ luaT_check_netbox_transport(struct lua_State *L, int idx)
 static int
 luaT_netbox_transport_gc(struct lua_State *L)
 {
+	printf("luaT_netbox_transport_gc\n");
 	struct netbox_transport *transport = luaT_check_netbox_transport(L, 1);
 	netbox_transport_destroy(transport);
 	return 0;
```
Before your commit I saw the message printed, but now I don't. Can it be a leak? Would it help if the weak callback would return true when has callback and false when it doesn't, and the transport in the latter case would stop self?

Reply:
You're totally right - the patch would result in a leak. Thanks for catching this. That's unfortunate that we didn't have a test case for this.

I don't like adding a return value to the callback, because that would complicate the code quite a bit (the callback is invoked from multiple places in `net_box.c`). Instead I turned `transport` into a weak reference. I also added a test case that ensures that both `transport` and `callback` are garbage collected.